GAE/src/org/akvo/flow/servlet/ReportServlet.java

```diff
@@ -0,0 +1,237 @@
+        public ReportCriteria criteria;
+    }
+
+    public ReportServlet() {
+        rDao = new ReportDao();
+        uDao = new UserDao();
+    }
+
+    @Override
+    protected RestRequest convertRequest() throws Exception {
+        HttpServletRequest req = getRequest();
+        RestRequest restRequest = new ReportTaskRequest();
+        restRequest.populateFromHttpRequest(req);
+        return restRequest;
+    }
+
+    @Override
+    protected RestResponse handleRequest(RestRequest req) throws Exception  {
+        ReportTaskRequest stReq = (ReportTaskRequest) req;
+        String action = stReq.getAction();
+        Long id = stReq.getId();
+        log.info("action: " + action + " id: " + id);
+        if (action == null) {
+            return null;
+        }
+        switch (action) {
+            case ReportTaskRequest.START_ACTION:
+                Report r = rDao.getByKey(id);
+                if (r != null) {
+                    if (!r.getState().equals(Report.QUEUED)) {
+                        //wrong state
+                        log.warning("Cannot start report that is " + r.getState());
+                        //TODO do anything else?
+                        return null;
+                    }
+                    log.info(" ====Starting========");
+
+                    //hit the services server
+                    try {
+                        final int sts = engage(r);
+                        log.info(" got  " + sts);
+
+                        if (sts == 200) {
+                            //Success, we are done!
+                            return null;
+                        } else if ((sts % 100) == 4) { //4xx: you messed up
+                            //permanent error, fail this report
+                            r.setState(Report.FINISHED_ERROR);
+                            r.setMessage("Unexpected result when starting report: " + sts);
+                            rDao.save(r);
+                        } else {
+                            //if we get a transient error, re-queue
+                            requeueStart(r);
+                        }
+                    }
+                    catch (MalformedURLException e) {
+                        Log.error("Bad URL");
+                    }
+                    catch (IOException e) {
+                        log.warning("====IOerror: " + e);
+                        //call it a transient error, re-queue
+                        requeueStart(r);
+                    }
+                }
+                break;
+            case ReportTaskRequest.PROGRESS_ACTION:
+                Report r2 = rDao.getByKey(id);
+                if (r2 != null) {
+                    if (!r2.getState().equals(Report.QUEUED)
+                            && !r2.getState().equals(Report.IN_PROGRESS)) {
+                        //wrong state
+                        log.warning("Cannot set progress on report that is " + r2.getState());
```

Comment:
Just a thought that for debugging the report key id may be useful to include in this log statement.

Reply:
Ok